File: lib/src/main/java/me/ameriod/lib/mvp/view/deligate/ControllerDelegateImpl.java
package me.ameriod.lib.mvp.view.deligate;

import android.os.Bundle;
import android.support.annotation.NonNull;

import me.ameriod.lib.mvp.Mvp;

public class ControllerDelegateImpl<V extends Mvp.View, P extends Mvp.Presenter<V>> implements ControllerDelegate<V, P>{

    private ControllerDelegateCallback<V, P> delegateCallback;
    private P presenter;

    public ControllerDelegateImpl(ControllerDelegateCallback<V, P> delegateCallback) {
        this.delegateCallback = delegateCallback;
    }

    @Override
    public void onCreateView() {
        if (presenter == null) {
            presenter = delegateCallback.createPresenter();
        }
        presenter.attachView(delegateCallback.getMvpView());
    }

    @Override
    public void onRestoreViewState(@NonNull Bundle savedViewState) {
        presenter.restoreState(savedViewState);
    }

    @Override
    public void onSaveViewState(@NonNull Bundle outState) {
        presenter.saveState(outState);
    }

    @Override
    public void onDestroyView() {
        presenter.detachView();
    }

    @Override
    public P getPresenter() {
        return presenter;
    }
}

File: lib/src/main/java/me/ameriod/lib/mvp/deligate/ControllerDelegateImpl.kt
package me.ameriod.lib.mvp.deligate

import android.os.Bundle

import me.ameriod.lib.mvp.Mvp

class ControllerDelegateImpl<V : Mvp.View, out P : Mvp.Presenter<V>>(private val delegateCallback:
                                                                     ControllerDelegateCallback<V, P>) : ControllerDelegate<V, P> {
    private var presenter: P? = null

    override fun onCreateView() {
        if (presenter == null) {
            presenter = delegateCallback.createPresenter()
        }
        presenter!!.attachView(delegateCallback.getMvpView())
    }

    override fun onRestoreViewState(savedViewState: Bundle) {
        presenter!!.restoreState(savedViewState)
    }

    override fun onSaveViewState(outState: Bundle) {
        presenter!!.saveState(outState)
    }

    override fun onDestroyView() {
        presenter!!.detachView()
    }

    override fun getPresenter(): P {
        return presenter!!
    }
}
